### DISCUSSION TITLE
How do you stub methods on a dataset?

### DISCUSSION THREAD

**Discussion by jcavalieri** [jeremyevans/sequel#1841] [General]
I'm using JRuby 9.3.3.0, Sequel 5.53.0, and jdbc-sqlite3 3.28.0.

I'm using mysql for our database which has the method `on_duplicate_key_update`. This, however, is not available on `jdbc-sqlite3` which is used with the unit tests. So I'm trying to stub the method with RSpec. Unfortunately I cannot because the Dataset object is frozen.

```
ArgumentError:
       Cannot proxy frozen objects, rspec-mocks relies on proxies for method stubbing and expectations.
```

I tried stubbing the constant `TRUE_FREEZE` to false. But that didn't work.

How would you go about stubbing a method on a Dataset object?

**Comment by jeremyevans** [comment]
Instead of trying to modify the frozen dataset, you create a modified copy of it using with_extend:

```ruby
ds = ds.with_extend{def on_duplicate_key_update(*args) self end}
```

Alternatively, if you want this method to be available for all of a Database's datasets:

```ruby
DB.extend_datasets{def on_duplicate_key_update(*args) self end}
```
